storage/client.go

```diff
@@ -0,0 +1,183 @@
+// * `settings` must be retained per instance
+// * `storageOption`s must be resolved in the order they are received
+// * all API errors must be wrapped in the gax-go APIError type
+// * any unimplemented interface methods must return a StorageUnimplementedErr
+//
+// TODO(noahdietz): This interface is currently not used in the production code
+// paths
+type storageClient interface {
+
+	// Top-level methods.
+
+	GetServiceAccount(ctx context.Context, project string, opts ...storageOption) (string, error)
+	CreateBucket(ctx context.Context, project string, attrs *BucketAttrs, opts ...storageOption) (*BucketAttrs, error)
+	ListBuckets(ctx context.Context, project string, opts ...storageOption) (BucketIterator, error)
+
+	// Bucket methods.
+
+	DeleteBucket(ctx context.Context, bucket string, conds *BucketConditions, opts ...storageOption) error
+	GetBucket(ctx context.Context, bucket string, conds *BucketConditions, opts ...storageOption) (*BucketAttrs, error)
+	UpdateBucket(ctx context.Context, uattrs *BucketAttrsToUpdate, conds *BucketConditions, opts ...storageOption) (*BucketAttrs, error)
+	LockBucketRetentionPolicy(ctx context.Context, bucket string, conds *BucketConditions, opts ...storageOption) error
+	ListObjects(ctx context.Context, bucket string, q *Query, opts ...storageOption) (*ObjectIterator, error)
+
+	// Object metadata methods.
+
+	DeleteObject(ctx context.Context, bucket, object string, conds *Conditions, opts ...storageOption) error
+	GetObject(ctx context.Context, bucket, object string, conds *Conditions, opts ...storageOption) (*ObjectAttrs, error)
+	UpdateObject(ctx context.Context, bucket, object string, uattrs *ObjectAttrsToUpdate, conds *Conditions, opts ...storageOption) (*ObjectAttrs, error)
+
+	// Default Object ACL methods.
+
+	DeleteDefaultObjectACL(ctx context.Context, bucket string, entity ACLEntity, opts ...storageOption) error
+	ListDefaultObjectACLs(ctx context.Context, bucket string, opts ...storageOption) ([]ACLRule, error)
+	UpdateDefaultObjectACL(ctx context.Context, opts ...storageOption) (ACLRule, error)
+
+	// Bucket ACL methods.
+
+	DeleteBucketACL(ctx context.Context, bucket string, entity ACLEntity, opts ...storageOption) error
+	ListBucketACLs(ctx context.Context, bucket string, opts ...storageOption) ([]ACLRule, error)
+	UpdateBucketACL(ctx context.Context, bucket string, entity ACLEntity, role ACLRole, opts ...storageOption) (ACLRule, error)
+
+	// Object ACL methods.
+
+	DeleteObjectACL(ctx context.Context, bucket, object string, entity ACLEntity, opts ...storageOption) error
+	ListObjectACLs(ctx context.Context, bucket, object string, opts ...storageOption) ([]ACLRule, error)
+	UpdateObjectACL(ctx context.Context, bucket, object string, entity ACLEntity, role ACLRole, opts ...storageOption) (ACLRule, error)
+
+	// Media operations.
+
+	ComposeObject(ctx context.Context, req composeObjectRequest, opts ...storageOption) (*ObjectAttrs, error)
```

Comment:
Also, maybe better to pass a pointer to these new request structs rather than copying?

Reply:
Good catch, done.